### PR TITLE
feat: smart sessions abstractions 1

### PR DESCRIPTION
**Pull request by vr16x**
## PR-Codex overview
This PR focuses on enhancing the SDK's functionality by introducing a debug mode, improving the action policy system, and refining session management with new utilities and error handling.

### Detailed summary
- Added `isDebugMode` to various SDK components for enhanced logging.
- Introduced new functions for building session actions and policies.
- Updated error handling to use `console.error`.
- Modified ABI and constants to support new features.
- Refined batch instruction handling with simulation overrides.
- Enhanced session quote execution with signing and error handling.

> The following files were skipped due to too many changes: `src/sdk/modules/validators/smartSessions/toMultichainSmartSessions.test.ts`, `src/sdk/account/decorators/buildSessionAction.ts`, `src/sdk/modules/validators/smartSessions/decorators/mee/prepareForPermissions.ts`, `src/sdk/account/decorators/buildSessionAction.test.ts`, `src/sdk/clients/decorators/mee/getSessionQuote.test.ts`, `src/sdk/clients/decorators/mee/getSessionQuote.ts`

> ✨ Ask PR-Codex anything about this PR by commenting with `/codex {your question}`

### CODE REVIEW

File: src/sdk/account/decorators/buildAction.test.ts [hunk 1]
```diff
@@ -0,0 +1,356 @@
...
+      ]
+    })
+  })
+
+  it("Build transfer action with default unrestricted policy", async () => {
+    const [transferAction] = mcNexus.buildAction({
+      type: "transfer",
+      data: {
+        chainIds: [network.chain.id],
+        contractAddress: zeroAddress
```
Comment (filmakarov) [inline src/sdk/account/decorators/buildAction.test.ts]: should we use some mock address instead of zero address to distinguish if the contractAddress is consumed properly or not?

File: src/sdk/account/decorators/buildAction.test.ts [hunk 2]
```diff
@@ -0,0 +1,356 @@
...
+    expect(transferAction.actionPolicies[0].policy).to.eq(SUDO_POLICY_ADDRESS)
+  })
+
+  it("Build transfer action with custom policies", async () => {
+    const [transferAction] = mcNexus.buildAction({
+      type: "transfer",
+      data: {
+        chainIds: [network.chain.id],
+        contractAddress: zeroAddress,
+        policies: [{ type: "sudo" }, { type: "usageLimit", limit: 10n }]
```
Comment (filmakarov) [inline src/sdk/account/decorators/buildAction.test.ts]: I know this is just for demo purposes, but, jfyi, putting sudo policy and usageLimit policy together makes no sense
if there's some restrictive policy , sudo policy is not needed

File: src/sdk/account/decorators/buildAction.test.ts [hunk 3]
```diff
@@ -0,0 +1,356 @@
...
+    expect(approveAction.actionTargetSelector).to.eq(functionSignature)
+    expect(approveAction.chainId).to.eq(network.chain.id)
+    expect(approveAction.actionPolicies.length).to.eq(2)
+    expect(approveAction.actionPolicies[0].policy).to.eq(SUDO_POLICY_ADDRESS)
+    expect(approveAction.actionPolicies[1].policy).to.eq(
+      UNIVERSAL_ACTION_POLICY_ADDRESS
+    )
+  })
+
+  it("Build custom action with custom policies", async () => {
```
Comment (filmakarov) [inline src/sdk/account/decorators/buildAction.test.ts]: the policy is not custom here, right?

File: src/sdk/account/decorators/buildAction.test.ts [hunk 4]
```diff
@@ -0,0 +1,356 @@
...
+                condition: "greaterThan",
+                comparisonValue: 0n
+              }
+            ]
+          }
+        ]
+      }
+    })
+
+    expect(customActions.length).to.eq(2)
```
Comment (filmakarov) [inline src/sdk/account/decorators/buildAction.test.ts]: should we also check resulting chain ids?

File: src/sdk/account/decorators/buildActionPolicy.ts [hunk 5]
```diff
@@ -0,0 +1,271 @@
...
+   */
+  valueLimitPerUse?: bigint
+}
+
+/**
+ * Time Frame Policy parameters.
+ * - validAfter: Unix timestamp in seconds after which valid.
+ * - validUntil: Unix timestamp in seconds until which valid.
+ */
+export type BuildTimeFrameActionPolicy = {
```
Comment (filmakarov) [inline src/sdk/account/decorators/buildActionPolicy.ts]: I suggest adding ...Params here and below, so the type is not called with a verb, it is confusing

like `BuildTimeFrameActionPolicyParams`

File: src/sdk/account/decorators/buildAction.test.ts [hunk 6]
```diff
@@ -0,0 +1,356 @@
...
+      }
+    })
+
+    expect(customAction).toBeDefined()
+    expect(customAction.actionTarget).to.eq(zeroAddress)
+    expect(customAction.actionTargetSelector).to.eq(functionSignature)
+    expect(customAction.chainId).to.eq(network.chain.id)
+    expect(customAction.actionPolicies.length).to.eq(1)
+    expect(customAction.actionPolicies[0].policy).to.eq(
+      UNIVERSAL_ACTION_POLICY_ADDRESS
```
Comment (filmakarov) [inline src/sdk/account/decorators/buildAction.test.ts]: does it make sense to actually check that maxLimit has been assigned properly?

File: src/sdk/account/decorators/buildAction.ts [hunk 7]
```diff
@@ -0,0 +1,241 @@
...
+ * @property policies - (optional) Policy constraints for action validation/execution.
+ */
+type BaseBuildActionParams = {
+  chainIds: number[]
+  contractAddress: Address
+  policies?: BuildActionPolicyTypes[]
+}
+
+/** Build 'transfer' ERC20 action. */
+export type BuildTransferAction = {
```
Comment (filmakarov) [inline src/sdk/account/decorators/buildAction.ts]: same here, I suggest adding ...Params

File: src/sdk/account/decorators/buildAction.test.ts [hunk 8]
```diff
@@ -0,0 +1,356 @@
...
+    const [transferFromAction] = mcNexus.buildAction({
+      type: "transferFrom",
+      data: {
+        chainIds: [network.chain.id],
+        contractAddress: zeroAddress,
+        policies: [
+          { type: "sudo" },
+          {
+            type: "spendingLimits",
+            tokenLimits: [{ token: zeroAddress, limit: 1n }]
```
Comment (filmakarov) [inline src/sdk/account/decorators/buildAction.test.ts]: I think there's no need to specify `token` in `tokenLimits`
it is excess, and it is confusing. what if someone specifies `contractAddress: token1` and `token: token2` 
I think we can safely expect `token` param in `tokenLimits` is always same as `contractAddress` param for the transferFrom action

File: src/sdk/account/decorators/buildActionPolicy.test.ts [hunk 9]
```diff
@@ -0,0 +1,105 @@
...
+    const spendingLimitsPolicy = mcNexus.buildActionPolicy({
+      type: "spendingLimits",
+      tokenLimits: [{ token: testToken, limit }]
+    })
+
+    expect(spendingLimitsPolicy).toBeDefined()
+    expect(spendingLimitsPolicy.policy).to.eq(SPENDING_LIMITS_POLICY_ADDRESS)
+  })
+
+  it("Build universal policy", async () => {
```
Comment (filmakarov) [inline src/sdk/account/decorators/buildActionPolicy.test.ts:90]: should we add more unit tests for this to verify how rules were created, etc.
test some edge cases with arguments etc

I think AI can do this relatively quick and we are going to be sure that everything is created properly.

Also, I would suggest introducing more e2e/integrations tests (in a separate file) that display that the actions composed this way are consumed by SS and Policies contracts properly.

File: src/sdk/account/decorators/buildAction.test.ts [hunk 10]
```diff
@@ -0,0 +1,356 @@
...
+    expect(customAction).toBeDefined()
+    expect(customAction.actionTarget).to.eq(zeroAddress)
+    expect(customAction.actionTargetSelector).to.eq(functionSignature)
+    expect(customAction.chainId).to.eq(network.chain.id)
+    expect(customAction.actionPolicies.length).to.eq(2)
+    expect(customAction.actionPolicies[0].policy).to.eq(
+      UNIVERSAL_ACTION_POLICY_ADDRESS
+    )
+    expect(customAction.actionPolicies[1].policy).to.eq(
+      SPENDING_LIMITS_POLICY_ADDRESS
```
Comment (filmakarov) [inline src/sdk/account/decorators/buildAction.test.ts]: why two policies are assigned here?

File: src/sdk/account/decorators/buildAction.ts [hunk 11]
```diff
@@ -0,0 +1,241 @@
...
+          tokenLimits: [
+            {
+              token: data.contractAddress,
+              limit: limitPerAction
+            }
+          ]
+        })
+
+        policies.push(spendingLimitPolicy)
+      }
```
Comment (filmakarov) [inline src/sdk/account/decorators/buildAction.ts]: I'd say, in all the cases where we have recipient, we can use uni action policy
so we do not end up with two policies.
because the case when there's only limitPerAction specified, can also be covered by uni action policy

File: src/sdk/account/decorators/buildAction.test.ts [hunk 12]
```diff
@@ -0,0 +1,356 @@
...
+              {
+                calldataOffset: calldataArgument(1),
+                condition: "equal",
+                comparisonValue: 0n
+              }
+            ]
+          }
+        ]
+      }
+    })
```
Comment (filmakarov) [inline src/sdk/account/decorators/buildAction.test.ts]: Shouldn't we abstract it even more? similar to erc20 spending limit

like

```
const [approveAction] = mcNexus.buildAction({
      type: "approve",
      token: tokenAddress,
      recepient: someAddress,
      maxAmount: 1e18,
      chainIds: [network.chain.id],
})
```

`data` with policies can be provided as well if someone needs deeper confuguration

**Comment by filmakarov** [comment]
thank you Venk! I like this PR, it will make working with SS much friendly for devs

File: src/sdk/account/decorators/buildAction.test.ts (hunk 3 above)
Comment (vr16x) [reply src/sdk/account/decorators/buildAction.test.ts]: Nope, fixed the test naming

File: src/sdk/account/decorators/buildAction.test.ts (hunk 4 above)
Comment (vr16x) [reply src/sdk/account/decorators/buildAction.test.ts]: Done

File: src/sdk/account/decorators/buildActionPolicy.ts (hunk 5 above)
Comment (vr16x) [reply src/sdk/account/decorators/buildActionPolicy.ts]: Done

File: src/sdk/account/decorators/buildAction.test.ts (hunk 6 above)
Comment (vr16x) [reply src/sdk/account/decorators/buildAction.test.ts]: It is already tested in different test case where the tx simulations is performed with maxLimit restrict and fund value more than max limit.

So that covers these things. The reasons why I haven't covered those cases is, we're using rhinestone SDK under the hood, so it make sense to cover the functionality tests but no need to add tests for whether maxLimit is properly encoding with sessions lib.

File: src/sdk/account/decorators/buildActionPolicy.test.ts (hunk 9 above)
Comment (vr16x) [reply src/sdk/account/decorators/buildActionPolicy.test.ts:90]: Regarding the integration tests, its already covered in the getSessionQuote.test.ts and toMultichainSmartSessions.test.ts file where it attempts to test universal action policies and others

Comment (vr16x) [reply src/sdk/account/decorators/buildActionPolicy.test.ts:90]: `should we add more unit tests for this to verify how rules were created, etc.
test some edge cases with arguments etc`

I don't think so. Because, whatever the custom abstractions that we've added, the test case is covered. The internal encoding of the rules and etc... is coming from rhinestone itself.

File: src/sdk/account/decorators/buildAction.ts (hunk 11 above)
Comment (vr16x) [reply src/sdk/account/decorators/buildAction.ts]: I will replace this entirely with universal action policy

File: src/sdk/account/decorators/buildAction.test.ts (hunk 12 above)
Comment (vr16x) [reply src/sdk/account/decorators/buildAction.test.ts]: Hey yes, I will do that

File: src/sdk/account/decorators/buildAction.test.ts (hunk 8 above)
Comment (vr16x) [reply src/sdk/account/decorators/buildAction.test.ts]: Done

File: src/sdk/account/decorators/buildAction.test.ts (hunk 2 above)
Comment (vr16x) [reply src/sdk/account/decorators/buildAction.test.ts]: Understood, I will refactor this

File: src/sdk/account/decorators/buildAction.test.ts (hunk 10 above)
Comment (vr16x) [reply src/sdk/account/decorators/buildAction.test.ts]: Refactored this into a single universal policy

File: src/sdk/account/decorators/buildAction.ts (hunk 11 above)
Comment (vr16x) [reply src/sdk/account/decorators/buildAction.ts]: DOne

File: src/sdk/account/decorators/buildAction.test.ts (hunk 12 above)
Comment (vr16x) [reply src/sdk/account/decorators/buildAction.test.ts]: Support for all the policies are added for all these functions.
